sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -187,6 +187,9 @@ class Dataset[T] private[sql](
     }
   }
 
+  // Wrap analyzed logical plan with an analysis barrier so we won't traverse/resolve it again.
+  @transient private val planWithBarrier: LogicalPlan = AnalysisBarrier(logicalPlan)
```

Comment:
I like the above idea https://github.com/apache/spark/pull/17770/files#r115913503 by @cloud-fan after traversing all the usage of `DataSet.logicalPlan`. If caching is the only issue we have, we can override `canonicalized` to its children.